llvm/lib/CodeGen/StackProtector.cpp

```diff
@@ -520,17 +521,12 @@ bool StackProtector::InsertStackProtectors() {
     HasIRCheck = true;
 
     // If we're instrumenting a block with a tail call, the check has to be
-    // inserted before the call rather than between it and the return. The
-    // verifier guarantees that a tail call is either directly before the
-    // return or with a single correct bitcast of the return value in between so
-    // we don't need to worry about many situations here.
+    // inserted before the call rather than between it and the return.
     Instruction *Prev = CheckLoc->getPrevNonDebugInstruction();
-    if (Prev && isa<CallInst>(Prev) && cast<CallInst>(Prev)->isTailCall())
+    if (Prev && isa<CallInst>(Prev) &&
+        cast<CallInst>(Prev)->isTailCall() &&
+        isInTailCallPosition(*cast<CallInst>(Prev), *TM)) {
```

Comment:
Can we do something like:

```
if (auto *CI = dyn_cast_or_null<CallInst>(Prev))
  if (isInTailCallPosition(*CU, *TM))
```

This avoids the double conversion and is slightly easier to read.

Reply:
Great! Thanks.